### PR TITLE
Switch to pipenv

### PR DESCRIPTION
**Pull request by milanandreew**

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+[dev-packages]
+coverage = "*"
+coveralls = "*"
+docopt = "*"
+flake8-commas = "*"
+flake8-quotes = "*"
+flake8 = "*"
+idna = "*"
+mccabe = "*"
```
Comment (sushain97) [inline Pipfile]: Unfortunately, a number of these are also transitive dependencies.

File: Pipfile [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+mypy = "0.580"
+pep8-naming = "*"
+pycodestyle = "*"
+pyflakes = "*"
+typed-ast = "*"
+
+[requires]
```
Comment (sushain97) [inline Pipfile]: Please remove this section.

File: Pipfile [hunk 3]
```diff
@@ -0,0 +1,19 @@
+[[source]]
+url = "https://pypi.org/simple"
+verify_ssl = true
+name = "pypi"
+
+[packages]
+apertium-streamparser = "*"
+certifi = "*"
```
[the remark this answers is not in the document]
Comment (milanandreew) [reply Pipfile]: Sorry, my mistake.

File: Pipfile [hunk 4]
```diff
@@ -0,0 +1,29 @@
+[[source]]
+url = "https://pypi.org/simple"
+verify_ssl = true
+name = "pypi"
+
+[packages]
+apertium-streamparser = "*"
+chardet = "*"
+
```
Comment (sushain97) [inline Pipfile]: Extra empty line?

File: Pipfile [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+requests = "*"
+tornado = "*"
+urllib3 = "*"
```
Comment (sushain97) [inline Pipfile]: I believe this is also transitive.

File: Pipfile [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+commentjson = "*"
+
+[dev-packages]
+coverage = "*"
+coveralls = "*"
+docopt = "*"
```
Comment (sushain97) [inline Pipfile]: Transitive?

File: Pipfile [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+tornado = "*"
+commentjson = "*"
+
+[dev-packages]
+coverage = "*"
+coveralls = "*"
+flake8-commas = "*"
+flake8-quotes = "*"
+flake8 = "*"
+idna = "*"
```
Comment (sushain97) [inline Pipfile]: idna, mccabe, and typed-ast are transitive. I think pycodestyle and pyflakes are as well.

File: Pipfile [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+mccabe = "*"
+mypy = "*"
+pep8-naming = "*"
+pycodestyle = "*"
+pyflakes = "*"
+typed-ast = "*"
+
```
Comment (sushain97) [inline Pipfile]: Extra line?

**Comment by milanandreew** [comment]
OK, resolving style checker errors might take a while...
(It's the middle of the night in Germany, so good night). 
I'll do the linter issues tomorrow.
Afterwards the build should not fail

**Comment by milanandreew** [comment]
OK, seems to be some kind of version support error, let me just try something...

**Comment by milanandreew** [comment]
mypy in requirements.txt
"python_version >= '3.5' # we DO NOT want the typing backport"
anyone got an idea how to fix this?

**Comment by sushain97** [comment]
@milanandreew pipenv also supports markers like this one. e.g. see this marker in streamparser: https://github.com/apertium/streamparser/blob/master/Pipfile#L15

File: Pipfile [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+coveralls = "*"
+flake8-commas = "==2.0.0"
+flake8-quotes = "==0.14.0"
+flake8 = "==3.5.0"
+mypy = "*"
+pep8-naming = "*"
+pycodestyle = "*"
+pyflakes = "*"
+
+
```
Comment (sushain97) [inline Pipfile]: There seem to be extra empty lines here.

I believe `pyflakes` and `pycodestyle` are also transitive dependencies.

File: Pipfile [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+requests = "*"
+tornado = "*"
+commentjson = "*"
+
+[dev-packages]
+coverage = "*"
+coveralls = "*"
+flake8-commas = "==2.0.0"
+flake8-quotes = "==0.14.0"
+flake8 = "==3.5.0"
```
Comment (sushain97) [inline Pipfile]: Do all of these need to be pinned? And if some do, can we allow patch version bumps?

**Comment by milanandreew** [comment]
`$ NONPAIRS=/tmp/languages python -m unittest tests/test*.py
Traceback (most recent call last):
  File "/opt/python/3.4.6/lib/python3.4/runpy.py", line 170, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/python/3.4.6/lib/python3.4/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/python/3.4.6/lib/python3.4/unittest/__main__.py", line 18, in <module>
    main(module=None)
  File "/opt/python/3.4.6/lib/python3.4/unittest/main.py", line 92, in __init__
    self.parseArgs(argv)
  File "/opt/python/3.4.6/lib/python3.4/unittest/main.py", line 139, in parseArgs
    self.createTests()
  File "/opt/python/3.4.6/lib/python3.4/unittest/main.py", line 146, in createTests
    self.module)
  File "/opt/python/3.4.6/lib/python3.4/unittest/loader.py", line 157, in loadTestsFromNames
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/opt/python/3.4.6/lib/python3.4/unittest/loader.py", line 157, in <listcomp>
    suites = [self.loadTestsFromName(name, module) for name in names]
  File "/opt/python/3.4.6/lib/python3.4/unittest/loader.py", line 125, in loadTestsFromName
    parent, obj = obj, getattr(obj, part)
AttributeError: 'module' object has no attribute 'test'
`
I have no idea what has caused this error.

**Comment by milanandreew** [comment]
Until tomorrow.

**Comment by sushain97** [comment]
Please refrain from opening and closing this PR.

**Comment by sushain97** [comment]
You're likely running into this error: https://stackoverflow.com/questions/25575073/attributeerror-module-object-has-no-attribute-tests. I suggest SSHing into the Travis CI box and seeing whether you can successfully start APy. You can also install Python 3.4 locally (might not be able to duplicate the issue).

**Comment by milanandreew** [comment]
Sorry, I always click the "Close and Comment" button by accident.

**Comment by milanandreew** [comment]
VIM (No, i'm not referring to improved vi, VIM stands for "very important milestone"). 
TODO: check which dependency versions need to be pinned.

**Comment by milanandreew** [comment]
So the 3.5 build can't cope with an outdated pipfile lock.
And it seems like the 3.4 problem has something to do with tornado.

**Comment by milanandreew** [comment]
Here comes the tricky bit

**Comment by milanandreew** [comment]
enhancement complete

File: Pipfile [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+requests = "*"
+tornado = "==4.2.1"
+commentjson = "*"
+
+[dev-packages]
+coverage = "*"
+coveralls = "*"
+flake8-commas = "*"
+flake8-quotes = "*"
+flake8 = "==3.5.0"
```
Comment (sushain97) [inline Pipfile]: Is it possible for us to allow patch upgrades to this package? i.e. something like `^3.5.0`?

Comment (milanandreew) [reply Pipfile]: I think so, some style issues would have to be resolved though.

Comment (sushain97) [reply Pipfile]: Even on patch versions? If so, I guess we can just create an issue to update it.

Comment (milanandreew) [reply Pipfile]: Yep.

Comment (milanandreew) [reply Pipfile]: Or just ignore the style conventions mentioned.

**Pull request merged**